[rocisa] Add rocisa

KKyang opened the pull request:
This is a port from ``TensileInstruction`` to ``rocisa``, including fixes like
1. Incorrect datatypes passing into functions
2. inst_str are changed after ``PrettyPrint`` is called
3. Generating times uplift avg in 16%

New optimizations are planned to implement in C++ after this commit if possible.

``rocisa`` use the CMake build system instead of the ``virtualenv``. The behavior of the TensileLite changed a bit with only one extra line:

```
cmake -DTENSILE_BIN=Tensile -DDEVELOP_MODE=ON -S <path-to-tensilelite-dir> -B <build-dir>
```
The script will be created in the build folder and will be named in Tensile.bat or Tensile.sh depending on the platform. You can then run the script as:

```
Tensile.sh config.yaml ./
```
or
```
Tensile.bat config.yaml ./
```

The ``CMakeLists.txt`` also offers build mode (WIP), the Python arguments have to be set during the configuration stage.

```
cmake -DTENSILE_BIN=Tensile -DBIN_ARGS="config.yaml ./" ..
cmake --build .
```

This offers integration with other ``CMakeLists.txt`` which does not require extra scripts to build and run TensileLite in one step.

KKyang committed:
Add rocisa

* Add rocisa

* Remove RegisterContainer

* Remove Singleton TensileInstructions

* Remove Item in Base.py

* Remove Containers

* Remove Instructions and Code

and lots of bug fixes

* Use DummyItem to speedup the process

* Add functions to speed up typeid check and vector pops

* Replace bottleneck functions

* Remove countType in Item

davidd-amd commented on tensilelite/Tensile/bin/Tensile:
Please put the logic for installing rocisa into a setup.py so that users can `pip install rocisa` and remove these bash scripts see details here:

https://medium.com/@kapilanr2003/a-comprehensive-walkthrough-of-python-c-binding-creation-253a6d11c1ff

KKyang replied:
I added the setup.py but if we remove those bash scripts, it'll become two steps which means each time the developer has to do pip install then run tensile while developing.

KKyang replied:
This need to be discussed to see if it is a good idea to do so.

davidd-amd replied:
Tensile now depends on rocisa. It is a python dependency and the conventional way to install a python dependency is via pip install. This precludes the need to require wrapper bash script to call a python script. Further, we are hiding details that modify users environments in a script which isn't ideal.

KKyang replied:
I use cmake instead now with a whitelist for all binaries.

stellaraccident commented:
Hi folks - would you all mind if I took a stab at a design doc for how I think the build should layer? I've got a pretty clear view of it in my head, based on having done some similar things on different projects. 

I have a feeling that if we can all look at the end state, we can reach a quick agreement that is where we want to go and then execute a little more directly on getting there. 

This stuff has been causing me many hours of overhead trying to integrate, and it would be a better use of my time to help get the code/build into a better organized state vs just dealing with the fallout.

KKyang commented:
> Hi folks - would you all mind if I took a stab at a design doc for how I think the build should layer? I've got a pretty clear view of it in my head, based on having done some similar things on different projects.
> 
> I have a feeling that if we can all look at the end state, we can reach a quick agreement that is where we want to go and then execute a little more directly on getting there.
> 
> This stuff has been causing me many hours of overhead trying to integrate, and it would be a better use of my time to help get the code/build into a better organized state vs just dealing with the fallout.

I'm happy to know what you want and how it should be built and talk a bit more how does this impact your local environment. Currently the build flow follows what Tensile does. If you have a guidance of the build flow that will be even better.

stellaraccident commented:
> > Hi folks - would you all mind if I took a stab at a design doc for how I think the build should layer? I've got a pretty clear view of it in my head, based on having done some similar things on different projects.
> > I have a feeling that if we can all look at the end state, we can reach a quick agreement that is where we want to go and then execute a little more directly on getting there.
> > This stuff has been causing me many hours of overhead trying to integrate, and it would be a better use of my time to help get the code/build into a better organized state vs just dealing with the fallout.
> 
> I'm happy to know what you want and how it should be built and talk a bit more how does this impact your local environment. Currently the build flow follows what Tensile does. If you have a guidance of the build flow that will be even better.

I will take time this week to write down an analysis and recommendation. Right now, the fragility of the hipBLASLt build setup is the leading cause of eng time being spent to keep it integrated on my side. As you note, much of this is legacy that comes from the way tensile was put together (which we should fix) and the other part is, honestly, just low quality cmake/tool implementations that need attention and need to pass a higher bar. I can clearly see that has evolved from a lot of people over a long time, so I'm sure I'm not offending anyone specifically by calling this out.

I think we can align to fix both, but I would very much like to avoid increasing the fragility without a concrete, written plan for the state we are working to achieve.

KKyang commented:
> > > Hi folks - would you all mind if I took a stab at a design doc for how I think the build should layer? I've got a pretty clear view of it in my head, based on having done some similar things on different projects.
> > > I have a feeling that if we can all look at the end state, we can reach a quick agreement that is where we want to go and then execute a little more directly on getting there.
> > > This stuff has been causing me many hours of overhead trying to integrate, and it would be a better use of my time to help get the code/build into a better organized state vs just dealing with the fallout.
> > 
> > 
> > I'm happy to know what you want and how it should be built and talk a bit more how does this impact your local environment. Currently the build flow follows what Tensile does. If you have a guidance of the build flow that will be even better.
> 
> I will take time this week to write down an analysis and recommendation. Right now, the fragility of the hipBLASLt build setup is the leading cause of eng time being spent to keep it integrated on my side. As you note, much of this is legacy that comes from the way tensile was put together (which we should fix) and the other part is, honestly, just low quality cmake/tool implementations that need attention and need to pass a higher bar. I can clearly see that has evolved from a lot of people over a long time, so I'm sure I'm not offending anyone specifically by calling this out.
> 
> I think we can align to fix both, but I would very much like to avoid increasing the fragility without a concrete, written plan for the state we are working to achieve.

So what problem are you actually facing? We know this needs to be refactored but this should be a long-term work instead of sticking with this PR. This PR is a kernel optimization work. Let's find a way we both can accept instead of letting the optimizations being blocked for a very long time.
Did you meet problems when recognizing Python versions? Or the compiler issues? Most of the setup are inherit from the hipBLASLt CMakeLists.txt. It shouldn't cause that much trouble unless I missed something. Can you point it out? What is the recommendation if virtualenv is a must? Does this modification require multiple steps that cause trouble to the developers?
What's the exact ETA of your guideline?

stellaraccident commented:
> > > > Hi folks - would you all mind if I took a stab at a design doc for how I think the build should layer? I've got a pretty clear view of it in my head, based on having done some similar things on different projects.
> > > > I have a feeling that if we can all look at the end state, we can reach a quick agreement that is where we want to go and then execute a little more directly on getting there.
> > > > This stuff has been causing me many hours of overhead trying to integrate, and it would be a better use of my time to help get the code/build into a better organized state vs just dealing with the fallout.
> > > 
> > > 
> > > I'm happy to know what you want and how it should be built and talk a bit more how does this impact your local environment. Currently the build flow follows what Tensile does. If you have a guidance of the build flow that will be even better.
> > 
> > 
> > I will take time this week to write down an analysis and recommendation. Right now, the fragility of the hipBLASLt build setup is the leading cause of eng time being spent to keep it integrated on my side. As you note, much of this is legacy that comes from the way tensile was put together (which we should fix) and the other part is, honestly, just low quality cmake/tool implementations that need attention and need to pass a higher bar. I can clearly see that has evolved from a lot of people over a long time, so I'm sure I'm not offending anyone specifically by calling this out.
> > I think we can align to fix both, but I would very much like to avoid increasing the fragility without a concrete, written plan for the state we are working to achieve.
> 
> So what problem are you actually facing? We know this needs to be refactored but this should be a long-term work instead of sticking with this PR. This PR is a kernel optimization work. Let's find a way we both can accept instead of letting the optimizations being blocked for a very long time. Did you meet problems when recognizing Python versions? Or the compiler issues? Most of the setup are inherit from the hipBLASLt CMakeLists.txt. It shouldn't cause that much trouble unless I missed something. Can you point it out? What is the recommendation if virtualenv is a must? Does this modification require multiple steps that cause trouble to the developers? What's the exact ETA of your guideline?

I would be happy to accept the PR if it does not regress the organization of the build -- which is already strained nearly to the breaking point. I didn't review the original to completion, but just on a starting look at the high level points (and please read my comments about the right way to structure a patch that makes such a large scale change to the build system -- it should not be inlined with a 20k LOC logic patch):

* You cannot invoke `make` like you were -- and certainly not at configure time.
* The CMake code does not adhere to any standards that I am aware of.
* You can't just use absolute paths into `/tmp/rocisa` with an expectation of exclusive control.
* You can't just pass `-j 8` in isolation.
* You can't hardcode `-DCMAKE_CXX_COMPILER=$ROCM_PATH/bin/amdclang++` as that is incompatible with almost everything except your dev setup (I have had to patch around many cases of this exact kind of sloppiness).
* You should not be hard coding `-DCMAKE_BUILD_TYPE=Debug`
* You should not be using advanced bash trampolines like this to invoke a basic tool (it is incompatible with Windows and it shouldn't be necessary)
* Your bash scripts do not enable error trapping and will malfunction silently on error (I have had to work around many cases of this in hipBLASLt already and it costs hours each time because so few things handle errors properly that it takes time to isolate what the root cause was).
* I couldn't tell you what this means `cd - > /dev/null || exit` except that I can tell you that it is wrong (you should never be exiting with a success error code on a tool like this)
* You should not have a self modifying import like this (as the Python loader can/will cache intermediates and fail in certain hard to debug situations):

```
try:
    from Tensile import TensileCreateLibrary
except ImportError:
    import os.path
    import sys
    parentdir = os.path.normpath(os.path.join(os.path.dirname(os.path.realpath(__file__)), "..", ".."))
    print(parentdir)
    sys.path.append(parentdir)

    from Tensile import TensileCreateLibrary
```

* You should not be tunneling different parts of the CMake setup together with `$ENV{PYTHONPATH}` as in `set(CommandLine PYTHONPATH=$ENV{PYTHONPATH} ${VIRTUALENV_BIN_DIR}/${VIRTUALENV_PYTHON_EXENAME} ${Script} ${Options} ${Tensile_LOGIC_PATH} ${Tensile_OUTPUT_PATH} HIP)`
* The procedure for finding nanobind is error prone and has been the cause of issues on other projects.
* There is special handling needed for `PYTHON_EXECUTABLE` vs `Python3_EXECUTABLE` that we know from other projects.

What I would suggest if we would like to proceed quickly (FTR - writing up a design doc to do this is not a "long term" effort. It can be done quickly and will make execution faster) is to send a patch that *only* adds the rocisa Python project to hipBLASLt and sets it up with a small skeleton of code. We review that in a couple of cycles and we possibly get someone who has more experience with this to finish/land it. Then you do as much as fast as you want on the actual implementation from there.

But the reviewers of the original patch were correct: there are far more than a few basic issues here. I'm taking it on face value that adding a nanobind based rocisa project is what we want to do -- add the project skeleton. Get that locked in. Then work on the implementation. If you sent a patch that was just working on the project organization, I'd even be willing to take your first draft and rework it for you in case if this isn't something that you natively have a lot of experience at. However, if commingling it with such a large change of project organization plus implementation, your reviewers have no choice but to simply block your patch.

KKyang commented:
> I would be happy to accept the PR if it does not regress the organization of the build -- which is already strained nearly to the breaking point. I didn't review the original to completion, but just on a starting look at the high level points (and please read my comments about the right way to structure a patch that makes such a large scale change to the build system -- it should not be inlined with a 20k LOC logic patch):
> 
> * You cannot invoke `make` like you were -- and certainly not at configure time.
> * The CMake code does not adhere to any standards that I am aware of.
> * You can't just use absolute paths into `/tmp/rocisa` with an expectation of exclusive control.
> * You can't just pass `-j 8` in isolation.
> * You can't hardcode `-DCMAKE_CXX_COMPILER=$ROCM_PATH/bin/amdclang++` as that is incompatible with almost everything except your dev setup (I have had to patch around many cases of this exact kind of sloppiness).
> * You should not be hard coding `-DCMAKE_BUILD_TYPE=Debug`
> * You should not be using advanced bash trampolines like this to invoke a basic tool (it is incompatible with Windows and it shouldn't be necessary)
> * Your bash scripts do not enable error trapping and will malfunction silently on error (I have had to work around many cases of this in hipBLASLt already and it costs hours each time because so few things handle errors properly that it takes time to isolate what the root cause was).
> * I couldn't tell you what this means `cd - > /dev/null || exit` except that I can tell you that it is wrong (you should never be exiting with a success error code on a tool like this)
> * You should not have a self modifying import like this (as the Python loader can/will cache intermediates and fail in certain hard to debug situations):
> 
> ```
> try:
>     from Tensile import TensileCreateLibrary
> except ImportError:
>     import os.path
>     import sys
>     parentdir = os.path.normpath(os.path.join(os.path.dirname(os.path.realpath(__file__)), "..", ".."))
>     print(parentdir)
>     sys.path.append(parentdir)
> 
>     from Tensile import TensileCreateLibrary
> ```
> 
> * You should not be tunneling different parts of the CMake setup together with `$ENV{PYTHONPATH}` as in `set(CommandLine PYTHONPATH=$ENV{PYTHONPATH} ${VIRTUALENV_BIN_DIR}/${VIRTUALENV_PYTHON_EXENAME} ${Script} ${Options} ${Tensile_LOGIC_PATH} ${Tensile_OUTPUT_PATH} HIP)`
> * The procedure for finding nanobind is error prone and has been the cause of issues on other projects.
> * There is special handling needed for `PYTHON_EXECUTABLE` vs `Python3_EXECUTABLE` that we know from other projects.
> 
> What I would suggest if we would like to proceed quickly (FTR - writing up a design doc to do this is not a "long term" effort. It can be done quickly and will make execution faster) is to send a patch that _only_ adds the rocisa Python project to hipBLASLt and sets it up with a small skeleton of code. We review that in a couple of cycles and we possibly get someone who has more experience with this to finish/land it. Then you do as much as fast as you want on the actual implementation from there.
> 
> But the reviewers of the original patch were correct: there are far more than a few basic issues here. I'm taking it on face value that adding a nanobind based rocisa project is what we want to do -- add the project skeleton. Get that locked in. Then work on the implementation. If you sent a patch that was just working on the project organization, I'd even be willing to take your first draft and rework it for you in case if this isn't something that you natively have a lot of experience at. However, if commingling it with such a large change of project organization plus implementation, your reviewers have no choice but to simply block your patch.

Thanks! Let me explain about this
1. The make is fixed and changed to cmake --build
2. We'll think of a proper place to put the build directory, but if you mean the one in the shell script, then that is an internal tool for developers. We don't want to cause troubles to the existing developers with different usage.
3. ~~we'll try to use nproc as possible~~ fixed
4. I was told that ROCM_PATH is an env and must exist in any of the rocm installed environment. I checked it and this also only exists in the bash script. Unfortunately, ROCm does not support alternative. The tests passed not only in my dev env but also all the math CI, but math CI does not have WIndows tested.
5. The two bash scripts are developers developing scripts, so it's hard coded to Debug.
6. The bash script is used for internal developers instead of having to perform two steps each time when writing codes. The vision is great, but we can't have people doing pip install and Tensile yaml working directory all the time causing troubles. AFAIK, these two shell script files shouldn't be touched by the build system.
7. ~~I forgot to add error handling, sorry.~~ fixed
8. These aren't even related to my PR

  You should not have a self modifying import like this (as the Python loader can/will cache intermediates and fail in certain hard to debug situations):
  ```
  try:
      from Tensile import TensileCreateLibrary
  except ImportError:
      import os.path
      import sys
      parentdir = os.path.normpath(os.path.join(os.path.dirname(os.path.realpath(__file__)), "..", ".."))
      print(parentdir)
      sys.path.append(parentdir)
  
      from Tensile import TensileCreateLibrary
  ```
  You should not be tunneling different parts of the CMake setup together with ``$ENV{PYTHONPATH}`` as in ``set(CommandLine PYTHONPATH=$ENV{PYTHONPATH} ${VIRTUALENV_BIN_DIR}/${VIRTUALENV_PYTHON_EXENAME} ${Script} ${Options} ${Tensile_LOGIC_PATH} ${Tensile_OUTPUT_PATH} HIP)``

9. How nanobind is found is recommended by Nanobind's doc. https://nanobind.readthedocs.io/en/latest/building.html#finding-nanobind
10. I changed to use Python_EXECUTABLE already

stellaraccident commented:
> > I would be happy to accept the PR if it does not regress the organization of the build -- which is already strained nearly to the breaking point. I didn't review the original to completion, but just on a starting look at the high level points (and please read my comments about the right way to structure a patch that makes such a large scale change to the build system -- it should not be inlined with a 20k LOC logic patch):
> > 
> > * You cannot invoke `make` like you were -- and certainly not at configure time.
> > * The CMake code does not adhere to any standards that I am aware of.
> > * You can't just use absolute paths into `/tmp/rocisa` with an expectation of exclusive control.
> > * You can't just pass `-j 8` in isolation.
> > * You can't hardcode `-DCMAKE_CXX_COMPILER=$ROCM_PATH/bin/amdclang++` as that is incompatible with almost everything except your dev setup (I have had to patch around many cases of this exact kind of sloppiness).
> > * You should not be hard coding `-DCMAKE_BUILD_TYPE=Debug`
> > * You should not be using advanced bash trampolines like this to invoke a basic tool (it is incompatible with Windows and it shouldn't be necessary)
> > * Your bash scripts do not enable error trapping and will malfunction silently on error (I have had to work around many cases of this in hipBLASLt already and it costs hours each time because so few things handle errors properly that it takes time to isolate what the root cause was).
> > * I couldn't tell you what this means `cd - > /dev/null || exit` except that I can tell you that it is wrong (you should never be exiting with a success error code on a tool like this)
> > * You should not have a self modifying import like this (as the Python loader can/will cache intermediates and fail in certain hard to debug situations):
> > 
> > ```
> > try:
> >     from Tensile import TensileCreateLibrary
> > except ImportError:
> >     import os.path
> >     import sys
> >     parentdir = os.path.normpath(os.path.join(os.path.dirname(os.path.realpath(__file__)), "..", ".."))
> >     print(parentdir)
> >     sys.path.append(parentdir)
> > 
> >     from Tensile import TensileCreateLibrary
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * You should not be tunneling different parts of the CMake setup together with `$ENV{PYTHONPATH}` as in `set(CommandLine PYTHONPATH=$ENV{PYTHONPATH} ${VIRTUALENV_BIN_DIR}/${VIRTUALENV_PYTHON_EXENAME} ${Script} ${Options} ${Tensile_LOGIC_PATH} ${Tensile_OUTPUT_PATH} HIP)`
> > * The procedure for finding nanobind is error prone and has been the cause of issues on other projects.
> > * There is special handling needed for `PYTHON_EXECUTABLE` vs `Python3_EXECUTABLE` that we know from other projects.
> > 
> > What I would suggest if we would like to proceed quickly (FTR - writing up a design doc to do this is not a "long term" effort. It can be done quickly and will make execution faster) is to send a patch that _only_ adds the rocisa Python project to hipBLASLt and sets it up with a small skeleton of code. We review that in a couple of cycles and we possibly get someone who has more experience with this to finish/land it. Then you do as much as fast as you want on the actual implementation from there.
> > But the reviewers of the original patch were correct: there are far more than a few basic issues here. I'm taking it on face value that adding a nanobind based rocisa project is what we want to do -- add the project skeleton. Get that locked in. Then work on the implementation. If you sent a patch that was just working on the project organization, I'd even be willing to take your first draft and rework it for you in case if this isn't something that you natively have a lot of experience at. However, if commingling it with such a large change of project organization plus implementation, your reviewers have no choice but to simply block your patch.
> 
> Thanks! Let me explain about this
> 

Yang-Wen and I spoke directly to clear some of this up. A couple of comments inline.

> 1. The make is fixed and changed to cmake --build
> 2. We'll think of a proper place to put the build directory, but if you mean the one in the shell script, then that is an internal tool for developers. We don't want to cause troubles to the existing developers with different usage.

Ok, as long as it stays only for developers. I've had to clean up some of these bash scripts elsewhere in the project because they didn't have proper error handling and had hardcoded tools paths. Let's make sure they don't become load bearing. This kind of thing tends to get copy-pasted, and then someone downstream ends up spending time trying to figure out why the build isn't deterministic or errors aren't being caught.

> 3. ~we'll try to use nproc as possible~ fixed
> 4. I was told that ROCM_PATH is an env and must exist in any of the rocm installed environment. I checked it and this also only exists in the bash script. Unfortunately, ROCm does not support alternative. The tests passed not only in my dev env but also all the math CI, but math CI does not have WIndows tested.

It doesn't. Some of the old rbuild/scripts set it, but most things have been explicitly fixed to be ROCm path independent and work without reference to an explicit ROCM path.

> 5. The two bash scripts are developers developing scripts, so it's hard coded to Debug.
> 6. The bash script is used for internal developers instead of having to perform two steps each time when writing codes. The vision is great, but we can't have people doing pip install and Tensile yaml working directory all the time causing troubles. AFAIK, these two shell script files shouldn't be touched by the build system.

There is another way we use on a lot of projects - just CMake -> Python. No venv, no pip, no packages and you get edit/continue. That is what I wanted to put in the design doc. It's actually pretty easy to do if you know how, and I think it would be a large simplification to the project.

> 7. ~I forgot to add error handling, sorry.~ fixed

Thanks. I'm unfortunately a bit jaded because I've spent hours going through hipBLASLt code to find where some error condition wasn't set and it proceeded to do the wrong thing. Let's always be doubly diligent about error handling...

> 8. These aren't even related to my PR
> 
> You should not have a self modifying import like this (as the Python loader can/will cache intermediates and fail in certain hard to debug situations):
> 
> ```
> try:
>     from Tensile import TensileCreateLibrary
> except ImportError:
>     import os.path
>     import sys
>     parentdir = os.path.normpath(os.path.join(os.path.dirname(os.path.realpath(__file__)), "..", ".."))
>     print(parentdir)
>     sys.path.append(parentdir)
> 
>     from Tensile import TensileCreateLibrary
> ```
> 
> You should not be tunneling different parts of the CMake setup together with `$ENV{PYTHONPATH}` as in `set(CommandLine PYTHONPATH=$ENV{PYTHONPATH} ${VIRTUALENV_BIN_DIR}/${VIRTUALENV_PYTHON_EXENAME} ${Script} ${Options} ${Tensile_LOGIC_PATH} ${Tensile_OUTPUT_PATH} HIP)`
> 
> 9. How nanobind is found is recommended by Nanobind's doc. https://nanobind.readthedocs.io/en/latest/building.html#finding-nanobind

Yeah, in practice, that has a variety of issues, and most projects I've worked on have had to adopt a more robust way to find it. We can leave it for now, but this is a frequent cause of problems, ime.

> 10. I changed to use Python_EXECUTABLE already

Thanks. Note that most things use FindPython3 (Python3_EXECUTABLE), but nanobind uses FindPython (Python_EXECUTABLE). We've typically had to do something like this in the top-most CMake file to make sure they don't get out of sync and end up accidentally mixing both: https://github.com/iree-org/iree/blob/35369313def343fb07f37a8ee5e53bed44ae62cf/CMakeLists.txt#L652

stellaraccident commented:
> > I would be happy to accept the PR if it does not regress the organization of the build -- which is already strained nearly to the breaking point. I didn't review the original to completion, but just on a starting look at the high level points (and please read my comments about the right way to structure a patch that makes such a large scale change to the build system -- it should not be inlined with a 20k LOC logic patch):
> > 
> > * You cannot invoke `make` like you were -- and certainly not at configure time.
> > * The CMake code does not adhere to any standards that I am aware of.
> > * You can't just use absolute paths into `/tmp/rocisa` with an expectation of exclusive control.
> > * You can't just pass `-j 8` in isolation.
> > * You can't hardcode `-DCMAKE_CXX_COMPILER=$ROCM_PATH/bin/amdclang++` as that is incompatible with almost everything except your dev setup (I have had to patch around many cases of this exact kind of sloppiness).
> > * You should not be hard coding `-DCMAKE_BUILD_TYPE=Debug`
> > * You should not be using advanced bash trampolines like this to invoke a basic tool (it is incompatible with Windows and it shouldn't be necessary)
> > * Your bash scripts do not enable error trapping and will malfunction silently on error (I have had to work around many cases of this in hipBLASLt already and it costs hours each time because so few things handle errors properly that it takes time to isolate what the root cause was).
> > * I couldn't tell you what this means `cd - > /dev/null || exit` except that I can tell you that it is wrong (you should never be exiting with a success error code on a tool like this)
> > * You should not have a self modifying import like this (as the Python loader can/will cache intermediates and fail in certain hard to debug situations):
> > 
> > ```
> > try:
> >     from Tensile import TensileCreateLibrary
> > except ImportError:
> >     import os.path
> >     import sys
> >     parentdir = os.path.normpath(os.path.join(os.path.dirname(os.path.realpath(__file__)), "..", ".."))
> >     print(parentdir)
> >     sys.path.append(parentdir)
> > 
> >     from Tensile import TensileCreateLibrary
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * You should not be tunneling different parts of the CMake setup together with `$ENV{PYTHONPATH}` as in `set(CommandLine PYTHONPATH=$ENV{PYTHONPATH} ${VIRTUALENV_BIN_DIR}/${VIRTUALENV_PYTHON_EXENAME} ${Script} ${Options} ${Tensile_LOGIC_PATH} ${Tensile_OUTPUT_PATH} HIP)`
> > * The procedure for finding nanobind is error prone and has been the cause of issues on other projects.
> > * There is special handling needed for `PYTHON_EXECUTABLE` vs `Python3_EXECUTABLE` that we know from other projects.
> > 
> > What I would suggest if we would like to proceed quickly (FTR - writing up a design doc to do this is not a "long term" effort. It can be done quickly and will make execution faster) is to send a patch that _only_ adds the rocisa Python project to hipBLASLt and sets it up with a small skeleton of code. We review that in a couple of cycles and we possibly get someone who has more experience with this to finish/land it. Then you do as much as fast as you want on the actual implementation from there.
> > But the reviewers of the original patch were correct: there are far more than a few basic issues here. I'm taking it on face value that adding a nanobind based rocisa project is what we want to do -- add the project skeleton. Get that locked in. Then work on the implementation. If you sent a patch that was just working on the project organization, I'd even be willing to take your first draft and rework it for you in case if this isn't something that you natively have a lot of experience at. However, if commingling it with such a large change of project organization plus implementation, your reviewers have no choice but to simply block your patch.
> 
> Thanks! Let me explain about this
> 
> 1. The make is fixed and changed to cmake --build
> 2. We'll think of a proper place to put the build directory, but if you mean the one in the shell script, then that is an internal tool for developers. We don't want to cause troubles to the existing developers with different usage.
> 3. ~we'll try to use nproc as possible~ fixed
> 4. I was told that ROCM_PATH is an env and must exist in any of the rocm installed environment. I checked it and this also only exists in the bash script. Unfortunately, ROCm does not support alternative. The tests passed not only in my dev env but also all the math CI, but math CI does not have WIndows tested.
> 5. The two bash scripts are developers developing scripts, so it's hard coded to Debug.
> 6. The bash script is used for internal developers instead of having to perform two steps each time when writing codes. The vision is great, but we can't have people doing pip install and Tensile yaml working directory all the time causing troubles. AFAIK, these two shell script files shouldn't be touched by the build system.
> 7. ~I forgot to add error handling, sorry.~ fixed
> 8. These aren't even related to my PR
> 
> You should not have a self modifying import like this (as the Python loader can/will cache intermediates and fail in certain hard to debug situations):
> 
> ```
> try:
>     from Tensile import TensileCreateLibrary
> except ImportError:
>     import os.path
>     import sys
>     parentdir = os.path.normpath(os.path.join(os.path.dirname(os.path.realpath(__file__)), "..", ".."))
>     print(parentdir)
>     sys.path.append(parentdir)
> 
>     from Tensile import TensileCreateLibrary
> ```
> 
> You should not be tunneling different parts of the CMake setup together with `$ENV{PYTHONPATH}` as in `set(CommandLine PYTHONPATH=$ENV{PYTHONPATH} ${VIRTUALENV_BIN_DIR}/${VIRTUALENV_PYTHON_EXENAME} ${Script} ${Options} ${Tensile_LOGIC_PATH} ${Tensile_OUTPUT_PATH} HIP)`
> 
> 9. How nanobind is found is recommended by Nanobind's doc. https://nanobind.readthedocs.io/en/latest/building.html#finding-nanobind
> 10. I changed to use Python_EXECUTABLE already

I think that if you really need to do the sub CMake invocation, the least bad way to do it is:

* Pass `${CMAKE_GENERATOR}` as the `-G` option to the cmake configure call, so that the sub-invocation uses the same generator as the main one.
* In your execute_process, add an env var `CMAKE_BUILD_PARALLEL_LEVEL=${NCORES}` (NCORES will be populated if you have prior done a `include(ProcessorCount)`.
* Typically such an env var is set by prefixing your command line with: `${CMAKE_COMMAND} -E env CMAKE_BUILD_PARALLEL_LEVEL=${NCORES} --`
* This way generators that default to single task (make/msbuild), will get it from the environment when `cmake --build` invokes them. And those that do the right thing already won't get tripped up by the extra flag.

stellaraccident commented on tensilelite/Tensile/cmake/TensileConfig.cmake:
```diff
@@ -209,7 +209,7 @@ function(TensileCreateLibraryFiles
   endif()
 
-  set(CommandLine ${VIRTUALENV_BIN_DIR}/${VIRTUALENV_PYTHON_EXENAME} ${Script} ${Options} ${Tensile_LOGIC_PATH} ${Tensile_OUTPUT_PATH} HIP)
+  set(CommandLine PYTHONPATH=$ENV{PYTHONPATH} ${VIRTUALENV_BIN_DIR}/${VIRTUALENV_PYTHON_EXENAME} ${Script} ${Options} ${Tensile_LOGIC_PATH} ${Tensile_OUTPUT_PATH} HIP)
```
Prefer `${CMAKE_COMMAND} -E env PYTHONPATH=$ENV{PYTHONPATH} -- <<rest of command line>>`

That is more portable.

KKyang replied:
Fixed

stellaraccident commented on tensilelite/Tensile/cmake/TensileConfig.cmake:
```diff
@@ -275,7 +275,7 @@ function(TensileCreateExtOpLibraries OutputFolder ArchStr)
     COMMAND ${CMAKE_COMMAND} -E make_directory ${build_tmp_dir}
     COMMAND ${CMAKE_COMMAND} -E make_directory ${OutputFolder}
-    COMMAND bash "${script}" "\"${Archs}\"" "${build_tmp_dir}" "${VIRTUALENV_HOME_DIR}" "${Tensile_BUILD_ID}"
+    COMMAND PYTHONPATH=$ENV{PYTHONPATH} bash "${script}" "\"${Archs}\"" "${build_tmp_dir}" "${VIRTUALENV_HOME_DIR}" "${Tensile_BUILD_ID}"
```
Ditto, set environment variables in command execution by prefixing with `${CMAKE_COMMAND} -E env ... --` (note that trailing double dash separates the remainder of the command line.

KKyang replied:
Fixed

stellaraccident commented on tensilelite/rocisa/CMakeLists.txt:
```diff
@@ -0,0 +1,77 @@
+set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
+set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
+
+if(DEFINED Python_EXECUTABLE AND Python_EXECUTABLE)
```
You really want to be relying on `find_package(Python)` for this kind of thing. It is not recommended to be manipulating the executable in this way.

Note that Python_EXECUTABLE and Python3_EXECUTABLE are not the same thing (for historic reasons). Nanobind uses the former. Many other things use the latter. We've taken to configuring both in the top-level CMake for a project: https://github.com/iree-org/iree/blob/main/CMakeLists.txt#L652

KKyang replied:
Yeah, that's the original version until the virtual env breaks the build while some of the dockers installed Python under /opt

KKyang replied:
Fixed with only showing messages if ``Python_EXECUTABLE`` is set.

stellaraccident commented on tensilelite/rocisa/CMakeLists.txt:
```diff
@@ -0,0 +1,77 @@
+# Use cmake -Dpybind11_DIR=$(python3 -m nanobind --cmakedir) .. to set pybind dir
+# Use cmake -DCMAKE_PREFIX_PATH=$(python3 -m nanobind --cmake_dir) .. to set nanobind
+
+find_package(Python COMPONENTS Interpreter Development REQUIRED)
```
You want Development.Module, not Development. There are many Python installations that are not set up for `Development` (which implies the ability to embed a Python interpreter and other things), whereas `Development.Module` supports only what is needed to compile extensions (which is pretty much always enabled).

Also, you really want to move this to the top before your first use of `Python_EXECUTABLE` (this is the thing which ensures that is correctly populated, and you are using it above).

KKyang replied:
the find_package has to be written in this exact way or else nanobind will return an error.

stellaraccident replied:
I'm pretty sure that is not the root cause. But ok. This will fail as written for some variations of python, but I guess can figure that out later. This stuff is fiddly.

KKyang replied:
Seems like it also works with ``Development.Module``. Thanks.

stellaraccident commented on tensilelite/rocisa/CMakeLists.txt:
```diff
@@ -0,0 +1,77 @@
+endif()
+
+# find nanobind
+execute_process(
```
Nit: It is possible that a super-project has already found nanobind. Typically you will condition all of this on `if(NOT nanobind_FOUND)` so that different project layers don't step on each other.

KKyang replied:
Fixed

stellaraccident commented on tensilelite/rocisa/cmake/RocisaConfig.cmake:
```diff
@@ -0,0 +1,74 @@
+    # Compute the installation prefix relative to this file.
+    get_filename_component(Rocisa_PREFIX "${CMAKE_CURRENT_LIST_FILE}" PATH)
+    get_filename_component(Rocisa_PREFIX "${Rocisa_PREFIX}" PATH)
+    if (WIN32)
```
Just change it to `"${Rocisa_PREFIX}/bin/RocisaGetPath${CMAKE_EXECUTABLE_SUFFIX}" and remove the redundant branch and code.

KKyang replied:
Fixed

stellaraccident commented on tensilelite/rocisa/cmake/RocisaConfig.cmake:
```diff
@@ -0,0 +1,74 @@
+  RESULT_VARIABLE result
+)
+if(NOT ${result} EQUAL 0)
+  message(STATUS "Output: ${output}")
```
message(FATAL_ERROR... please.

KKyang replied:
Fixed

stellaraccident commented on tensilelite/rocisa/cmake/RocisaConfig.cmake:
```diff
@@ -0,0 +1,74 @@
+  message(STATUS "rocIsa configuring done.")
+endif()
+execute_process(
+  COMMAND nproc
```
```
include(ProcessorCount)
# Then reference with ${NPROC}
```

KKyang replied:
Fixed

stellaraccident commented on tensilelite/rocisa/cmake/RocisaConfig.cmake:
```diff
@@ -0,0 +1,74 @@
+  OUTPUT_STRIP_TRAILING_WHITESPACE
+)
+execute_process(
+  COMMAND ${CMAKE_COMMAND} --build . -j ${NPROC_COUNT}
```
`COMMAND ${CMAKE_COMMAND} -E env CMAKE_BUILD_PARALLEL_LEVEL=${NPROC} -- ${CMAKE_COMMAND} --build .`

KKyang replied:
Fixed

stellaraccident commented on tensilelite/rocisa/cmake/RocisaConfig.cmake:
```diff
@@ -0,0 +1,74 @@
+  message(FATAL_ERROR "Result: ${result}")
+else()
+  message(STATUS "rocIsa building done.")
+  set(ENV{PYTHONPATH} "${ROCISA_VIRTUAL_DIR}/build/lib:$ENV{PYTHONPATH}")
```
A note on what is consuming this PYTHONPATH please. It is very irregular to be communicating by env var within a CMake configure session like this.

KKyang replied:
I use ``Rocisa_LIBRARIES`` instead.

stellaraccident commented on tensilelite/rocisa/cmake/RocisaConfig.cmake:
```diff
@@ -0,0 +1,74 @@
+file(MAKE_DIRECTORY ${ROCISA_VIRTUAL_DIR}/build)
+execute_process(
+  COMMAND ${CMAKE_COMMAND} -DPython_EXECUTABLE=${VIRTUALENV_BIN_DIR}/${VIRTUALENV_PYTHON_EXENAME} ..
+  WORKING_DIRECTORY ${ROCISA_VIRTUAL_DIR}/build
```
I'm having a bit of trouble following the paths, but this looks like you are creating the `build` directory under the source dir if I am following. Can you please not do that? If you need a place to stash it, at least use something like `${PROJECT_BINARY_DIR}/rocisa_build` (or something -- so long as it is explicitly in the build tree).

KKyang replied:
Fixed

stellaraccident commented on tensilelite/MANIFEST.in:
```diff
@@ -1,3 +1,5 @@
+recursive-include rocisa *.hpp *.cpp *.txt *.cmake
+recursive-exclude rocisa/build *.txt
```
I assume that this exclude here is because you are creating the build directory in the source dir. If you stop doing that, you should remove this exclude.

KKyang replied:
``pip install .`` does create a build folder, adding this in case someone accidentally use this command and the CMakeCache.txt is copied

stellaraccident replied:
Ok.

stellaraccident commented on tensilelite/Tensile/bin/Tensile.sh:
```diff
@@ -0,0 +1,70 @@
+# SOFTWARE.
+#
+################################################################################
+
```
`set -e`

(or `set -euo pipefail` for the most robustness)

KKyang replied:
fixed

stellaraccident commented on tensilelite/Tensile/bin/Tensile.sh:
```diff
@@ -0,0 +1,70 @@
+    echo $absolute_path
+}
+
+ROCM_PATH="${ROCM_PATH}"
```
`ROCM_PATH="${ROCM_PATH:-/opt/rocm}"`

KKyang replied:
Fixed

stellaraccident commented on tensilelite/Tensile/bin/TensileCreateLibrary.sh:
```diff
@@ -0,0 +1,70 @@
+# SOFTWARE.
+#
+################################################################################
+
```
set -e

(or set -euo pipefail for the most robustness)

KKyang replied:
Fixed

stellaraccident commented on tensilelite/Tensile/bin/TensileCreateLibrary.sh:
```diff
@@ -0,0 +1,70 @@
+    echo $absolute_path
+}
+
+ROCM_PATH="${ROCM_PATH}"
```
ROCM_PATH="${ROCM_PATH:-/opt/rocm}"

KKyang replied:
Fixed

stellaraccident commented on tensilelite/Tensile/bin/TensileCreateLibrary.sh:
```diff
@@ -0,0 +1,70 @@
+### Build rocisa
+# Currently the path to rocisa is fixed
+mkdir -p $WORKING_DIR/rocisa
+return_code=$?
```
If you use `set -e` mode, the script will fail if any command fails, which is usually what you want and can avoid explicit handling.

KKyang replied:
Fixed

stellaraccident commented on CMakeLists.txt:
```diff
@@ -236,6 +236,7 @@ else()
     endif()
     message(STATUS "Adding ${VIRTUALENV_HOME_DIR} to CMAKE_PREFIX_PATH")
     list(APPEND CMAKE_PREFIX_PATH ${VIRTUALENV_HOME_DIR})
+    find_package(Rocisa 0.1.0 EXACT REQUIRED HIP LLVM PATHS "${INSTALLED_TENSILE_PATH}")
```
I'm having trouble figuring out what this means. Specifically, I'm not sure why the "HIP LLVM" part is here and I suspect that is being ignored by CMake.

KKyang replied:
Sorry, this is copied from the Tensile find_package below. Will remove it, thanks!

stellaraccident replied:
It can be a bit fiddly to use find_package for a fixed PATH that should not be found anywhere else.

Recommend something like this: `find_package(Rocisa 0.1.0 EXACT REQUIRED CONFIG NO_DEFAULT_PATH PATHS "${INSTALLED_TENSILE_PATH}")`

(without `CONFIG`, cmake uses a more elaborate legacy procedure, and in both cases, iirc some system level things are checked before the PATHS you provide unless if you specify NO_DEFAULT_PATH)

KKyang replied:
I found out that the PATH is the one which is not needed. I changed to ``EXACT REQUIRED)`` only

KKyang commented on tensilelite/Tensile/cmake/TensileConfig.cmake:
```diff
@@ -24,6 +24,10 @@
+if(NOT Rocisa_FOUND)
+    find_package(Rocisa 0.1.0 EXACT REQUIRED)
+endif()
+
```
@stellaraccident I moved the ``find_package`` here since only Tensile needs it. The env is passed out using ``Rocisa_LIBRARIES``.

davidd-amd commented:
> I think that if you really need to do the sub CMake invocation, the least bad way to do it is:

@stellaraccident thanks for the sage CMake advice. I wanted to offer another perspective on having cmake invoke cmake. As you said this is the least bad way to do things implying it isn't ideal in any scenario. Tensilelite now depends on the rocisa python module in the same way that we depend on joblib. it seems to me like this dependency should be installed into our python environment as we would any other python package.

On another note, I've had a proof of principal build system update up for a while now that would removes the pyenv mechanism altogether (I believe it was a starting point for a patch for TheRock). I would like to begin polishing that PR and open it for review so that we can continue these discussions in the context of a PR scoped to the CMake support.

KKyang commented:
> > I think that if you really need to do the sub CMake invocation, the least bad way to do it is:
> 
> @stellaraccident thanks for the sage CMake advice. I wanted to offer another perspective on having cmake invoke cmake. As you said this is the least bad way to do things implying it isn't ideal in any scenario. Tensilelite now depends on the rocisa python module in the same way that we depend on joblib. it seems to me like this dependency should be installed into our python environment as we would any other python package.
> 
> On another note, I've had a proof of principal build system update up for a while now that would removes the pyenv mechanism altogether (I believe it was a starting point for a patch for TheRock). I would like to begin polishing that PR and open it for review so that we can continue these discussions in the context of a PR scoped to the CMake support.

I don't think this is a good idea to just ``pip install`` since joblib is a complete module while rocisa will be modified at any time. This not only screw up the docker environment when switching between branches or commits (you have to really take care with version you've installed) but also cause confusions to other users outside the team.

stellaraccident commented:
> > > I think that if you really need to do the sub CMake invocation, the least bad way to do it is:
> > 
> > 
> > @stellaraccident thanks for the sage CMake advice. I wanted to offer another perspective on having cmake invoke cmake. As you said this is the least bad way to do things implying it isn't ideal in any scenario. Tensilelite now depends on the rocisa python module in the same way that we depend on joblib. it seems to me like this dependency should be installed into our python environment as we would any other python package.
> > On another note, I've had a proof of principal build system update up for a while now that would removes the pyenv mechanism altogether (I believe it was a starting point for a patch for TheRock). I would like to begin polishing that PR and open it for review so that we can continue these discussions in the context of a PR scoped to the CMake support.
> 
> I don't think this is a good idea to just `pip install` since joblib is a complete module while rocisa will be modified at any time. This not only screw up the docker environment when switching between branches or commits (you have to really take care with version you've installed) but also cause confusions to other users outside the team.

Let's start a design doc with an alternatives considered section that we can collaborate on. I agree with @KKyang's concerns about the development process and integration implications of making the development flow include external management of a pyenv. Deps external to the project should go in a requirements file and the user is responsible for populating them (with us recommending to use a virtual env). But artifacts that the project builds need to be integral to the build, not separate from it. 

My alternative considered will not have any project managed pyenv or sub cmake build. We built llvm that way for the same reasons, and it is much more complicated use cases and has held up.

Do you want to start a hackmd? (Hackmd.io is what my team uses for public design docs)

b-shi commented on tensilelite/rocisa/rocisa/src/container.cpp:
```diff
@@ -0,0 +1,533 @@
+        return createGPR("acc", name, regNum);
+    }
+
+    std::shared_ptr<RegisterContainer> mgpr(const Holder& holder, float regNum = 1.f)
```
Is this only used for `m0`? Could we make this a special op like what you did for `vcc` and `exec`?

KKyang replied:
``VCC`` and ``EXEC`` does not require any extra inputs, just simply use ``EXEC()`` and ``VCC()``. The behavior here is the same as Python.

b-shi replied:
The confusion is that the ISA does not have `mgpr`. It looks like we are just using `mgpr(0)` to mean `m0` (AFAIK there is no `m1`, `m2`,..).

> just simply use EXEC() and VCC()

Since `vcc`,`exec`, and `m0` are all scalar registers, could we define something like `M0()` instead (i,e `dst=M0()...`)?

KKyang replied:
It'll be a good idea; I'll add it to the TODO list. But will not change in this PR because it'll require to change the code in the generator as well. Currently the diffs in the generator are only to make the function works properly. None of the behavior has changed. I'll try to make the minimal changes else the PR will getting bigger and bigger.

b-shi commented on tensilelite/rocisa/rocisa/include/container.hpp:
```diff
@@ -0,0 +1,1086 @@
+            , offset(offset)
+            , offset0(offset0)
+            , offset1(offset1)
+            , gds(gds)
```
Just a comment. We probably don't need the `gds` field (not sure if its actually [supported](https://github.com/ROCm/ROCm/issues/1296#issuecomment-1338772042)..)

b-shi commented on tensilelite/rocisa/rocisa/src/instruction/common.cpp:
```diff
@@ -0,0 +1,1963 @@
+            return new rocisa::SAbsI32(self);
+        });
+
+    nb::class_<rocisa::SMaxI32, rocisa::CommonInstruction>(m_common, "SMaxI32")
```
The declarations here look boiler plate, could we use a macro to reduce the number of lines? Something like
```c++
#define ROCISA_DECLARE_DST_SRC(INSTRUCTION) \
  nb::class_<rocisa::INSTRUCTION, rocisa::CommonInstruction>(m_common, TOSTRING(INSTRUCTION)) \
  ... \
  .. return new rocisa::INSTRUCTION(self); });  

#define ROCISA_DECLARE_DST_SRC0_SRC1(INSTRUCTION) ..

ROCISA_DECLARE_DST_SRC(SAbsI32);
ROCISA_DECLARE_DST_SRC0_SRC1(SMaxI32);
ROCISA_DECLARE_DST_SRC0_SRC1(SMaxU32);
...
```
Similar for the declaration in headers, and for other types (cvt, instruction, mem, etc..)

KKyang replied:
It'll be a little complicated since there might be an exception. Or by that time we'll choose not to use macro instead? (Actually I'm not a fan of a macro because if compiler error occurs, the line happens in the macro)

b-shi replied:
For instructions where there is an exception to the declaration, we can just use not use the macro. I'm guessing a lot of these declarations will be similar though.

> if compiler error occurs, the line happens in the macro

Fair point. There are definitely trade-offs with using macros.

b-shi commented on tensilelite/rocisa/rocisa/include/code.hpp:
```diff
@@ -0,0 +1,1613 @@
+{
+    struct Label : public Item
+    {
+        std::variant<std::string, int> label;
```
For labels, could we internally prefix the name with `.L`, something like 
```
myLabel -> .LmyLabel
```
and use that in the final asm src? This would prevent the assembler from adding an entry in the symbol table for each label we use.

KKyang replied:
I misunderstood your meaning. Other teams need these symbols. We'll probably keep them now.
~~Seems not working in amdclang++~~

b-shi commented on tensilelite/rocisa/rocisa/include/code.hpp:
```diff
@@ -0,0 +1,1613 @@
+
+    union SrdUpperFields9XX
+    {
+        struct
```
Just a comment. Will these fields be copied one by one or with a single 32b copy when used to init the buffer descriptor? If doing a single copy, I'm not 100% sure the ordering of the bit fields is guaranteed to be respected by the compiler.

KKyang replied:
single 32b, and ordering is guaranteed

KKyang committed:
Use sub_directory instead

stellaraccident commented on CMakeLists.txt:
```diff
@@ -188,6 +188,9 @@ else()
     endif( )
 
+    # Add rocisa
+    add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/tensilelite/rocisa)
```
Style nit: typically leave off ${CMAKE_CURRENT_SOURCE_DIR} for add_subdirectory since it is the default.

stellaraccident commented on cmake/virtualenv.cmake:
```diff
@@ -35,6 +35,10 @@ function(virtualenv_install)
       COMMAND_ECHO STDOUT
     )
+    execute_process(
+      COMMAND ${VIRTUALENV_BIN_DIR}/${VIRTUALENV_PYTHON_EXENAME} -m pip install --upgrade nanobind
```
Nit for the future: should check error return on all of these and fail.

KKyang replied:
Thanks. Since I'm not using virtualenv anymore, I should remove this.

stellaraccident commented on tensilelite/rocisa/CMakeLists.txt:
```diff
@@ -0,0 +1,76 @@
+        RESULT_VARIABLE result
+    )
+    if(result)
+        message(FATAL_ERROR "Did you forget to install nanobind?")
```
Nit: I usually like being a bit more explicit in such errors. Include something like `(typically this is done with ${Python_EXECUTABLE} -m pip install nanobind)` at the end. Specifically making sure that the Python_EXECUTABLE is in the error message tends to help a lot when it fails in the field because the error messages you get in an issue will often have more information.

stellaraccident reviewed (dismissed):
This looks much improved to me. Thank you for taking the time on the build structure concerns. Much appreciated.

KKyang committed:
Use FetchContent

KKyang commented:
Use ``FetchContent`` so no ``pip install -r requirements.txt`` while building ``rocisa``.

KKyang commented:
Fixing ``VMovB32`` comments not using keys and cause errors after adding ``sdwa`` modifiers.

stellaraccident commented on tensilelite/rocisa/CMakeLists.txt:
```diff
@@ -36,25 +36,19 @@ endif()
 endif()
-message(STATUS "Nanobind path: ${nanobind_DIR}")
 
+include(FetchContent)
```
Add a todo: make fetching optional for hermetic builds.

This is a good way to get this patch landed, but we may want to condition it a bit more so that build environments that are hermetic have a path. Fine to work that out in a follow-up since this is already a big enough change.

KKyang committed:
Add DEVELOP_MODE for tensilelite CMakeLists.txt

KKyang commented:
1. Azure CI not working properly since last friday
2. gfx942 test won't start, local results
![image](https://github.com/user-attachments/assets/ee1e77c8-0249-4c62-b0fe-4385bfd1381e)

![image](https://github.com/user-attachments/assets/d51f2db3-8ea7-4f1c-99d1-c5b1bbeb28b2)

KKyang commented:
New tox CI added
https://github.com/ROCm/rocJenkins/pull/1070/
